src/pip/_internal/index.py

```diff
@@ -939,15 +939,19 @@ def _get_encoding_from_headers(headers):
     return None
 
 
-_CLEAN_LINK_RE = re.compile(r'[^a-z0-9$&+,/:;=?@.#%_\\|-]', re.I)
+_CLEAN_LINK_RE = re.compile(r"^(/[a-z])%3A", re.I)
 
 
 def _clean_link(url):
     # type: (str) -> str
     """Makes sure a link is fully encoded.  That is, if a ' ' shows up in
     the link, it will be rewritten to %20 (while not over-quoting
     % or other characters)."""
-    return _CLEAN_LINK_RE.sub(lambda match: '%%%2x' % ord(match.group(0)), url)
+    result = urllib_parse.urlparse(url)
+    return urllib_parse.urlunparse(
+        result._replace(
+            path=_CLEAN_LINK_RE.sub(r"\1:", urllib_parse.quote(
+                urllib_parse.unquote(result.path)))))
```

Comment:
This is a bit unwieldy IMO. I would split this into several short statements instead.

Reply:
Good point. I split it into several lines.